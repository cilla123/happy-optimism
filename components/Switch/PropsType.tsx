export default interface PropsType {
  theme?: 'primary' | 'success' | 'warning' | 'error';
  checked?: boolean;
  defaultChecked?: boolean;
  disabled?: boolean;
  onChange?: (checked: boolean) => void;
}
