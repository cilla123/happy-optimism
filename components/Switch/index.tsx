import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';

const getChecked = (props, defaultChecked) => {
  if ('checked' in props && props.checked) {
    return props.checked;
  }
  if ('defaultChecked' in props && props.defaultChecked) {
    return props.defaultChecked;
  }
  return defaultChecked;
};

export interface SwitchProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Switch extends PureComponent<SwitchProps, any> {

  static defaultProps = {
    prefixCls: 'za-switch',
    theme: 'primary',
    disabled: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      checked: getChecked(props, false),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  onValueChange = () => {
    const { disabled, onChange } = this.props;
    if (disabled) {
      return;
    }

    const checked = !this.state.checked;
    this.setState({ checked });
    if (typeof onChange === 'function') {
      onChange(checked);
    }
  }

  render() {
    const { prefixCls, className, theme, disabled } = this.props;
    const { checked } = this.state;

    const cls = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      checked,
      disabled,
    });

    return (
      <span className={cls}>
        <input
          type="checkbox"
          className={`${prefixCls}-input`}
          disabled={disabled}
          checked={checked}
          onChange={this.onValueChange}
        />
      </span>
    );
  }
}
