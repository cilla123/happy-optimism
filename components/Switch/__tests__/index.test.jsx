import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Switch from '../index';

describe('Switch', () => {
  it('renders correctly', () => {
    const wrapper = render(<Switch />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('defaultChecked', () => {
    const wrapper = render(<Switch defaultChecked />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('onChange', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Switch checked onChange={onChange} />);
    wrapper.find('input').simulate('change', { target: { checked: false } });
    expect(onChange).toBeCalledWith(false);
  });

  it('receive new checked', () => {
    const wrapper = shallow(<Switch className="test-class" />);
    wrapper.setProps({ checked: true });
  });
});
