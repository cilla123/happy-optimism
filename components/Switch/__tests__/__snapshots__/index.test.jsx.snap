// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch defaultChecked 1`] = `
<span
  class="za-switch theme-primary checked"
>
  <input
    checked=""
    class="za-switch-input"
    type="checkbox"
  />
</span>
`;

exports[`Switch renders correctly 1`] = `
<span
  class="za-switch theme-primary"
>
  <input
    class="za-switch-input"
    type="checkbox"
  />
</span>
`;
